### PR TITLE
Hoist the desired ReactDOM import expression

### PR DESCRIPTION
**Pull request by huozhi**
* hoist `react-dom/server` imports to reduce module load time in nodejs
* simplify `reactRoot` detection condition by checking the streaming rendering API we're using. (if it doesn't existed, like react 17, then we won't enable `reactRoot`)
* Merge `__NEXT_CONCURRENT_FEATURES` into `__NEXT_REACT_ROOT` env var since they're identical now

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: d7ddc6a4718be7aa3db25a2d9ec3de484f0133cc

`yarn testheadless test/integration/amphtml-custom-optimizer/test/index.test.js` 
- AMP Custom Optimizer > should build and start for static page
- AMP Custom Optimizer > should build and start for dynamic page

<details>
<summary>Expand output</summary>

  ● AMP Custom Optimizer › should build and start for static page

    command failed with code 1
    warn  - You have enabled experimental feature(s).
    warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

    info  - Checking validity of types...
    warn  - The Next.js plugin was not detected in your ESLint configuration. See https://nextjs.org/docs/basic-features/eslint#migrating-existing-config
    info  - Creating an optimized production build...
    info  - Using locally built binary of @next/swc
    info  - Compiled successfully
    info  - Collecting page data...
    info  - Generating static pages (0/3)

    Error occurred prerendering page "/". Read more: https://nextjs.org/docs/messages/prerender-error
    TypeError: ReactDOMServer.renderToReadableStream is not a function

      122 | }
      123 | async function renderToInitialStream({ ReactDOMServer , element  }) {
    > 124 |     return await ReactDOMServer.renderToReadableStream(element);
          |                                 ^
      125 | }
      126 | async function continueFromInitialStream({ suffix , dataStream , generateStaticHTML , flushEffectHandler , renderStream  }) {
      127 |     const closeTag = '</body></html>';

      at Object.renderToInitialStream (../packages/next/dist/server/node-web-streams-helper.js:124:33)
      at renderShell (../packages/next/dist/server/render.js:882:65)
      at Object.renderPage (../packages/next/dist/server/render.js:795:28)
      at Object.defaultGetInitialProps (../packages/next/dist/server/render.js:391:51)
      at Function.getInitialProps (integration/amphtml-custom-optimizer/.next/server/pages/_document.js:590:20)
      at Object.loadGetInitialProps (../packages/next/dist/shared/lib/utils.js:65:29)
      at documentInitialProps (../packages/next/dist/server/render.js:817:48)
      at renderDocument (../packages/next/dist/server/render.js:942:49)
      at renderToHTML (../packages/next/dist/server/render.js:971:34)
      at ../async /home/runner/work/next.js/next.js/packages/next/dist/export/worker.js:253:36
      Error occurred prerendering page "/404". Read more: https://nextjs.org/docs/messages/prerender-error
      TypeError: ReactDOMServer.renderToReadableStream is not a function
      at Object.renderToInitialStream (../packages/next/dist/server/node-web-streams-helper.js:124:33)
      at renderShell (../packages/next/dist/server/render.js:882:65)
      at Object.renderPage (../packages/next/dist/server/render.js:795:28)
      at Object.defaultGetInitialProps (../packages/next/dist/server/render.js:391:51)
      at Function.getInitialProps (integration/amphtml-custom-optimizer/.next/server/pages/_document.js:590:20)
      at Object.loadGetInitialProps (../packages/next/dist/shared/lib/utils.js:65:29)
      at documentInitialProps (../packages/next/dist/server/render.js:817:48)
      at renderDocument (../packages/next/dist/server/render.js:942:49)
      at renderToHTML (../packages/next/dist/server/render.js:971:34)
      at ../async /home/runner/work/next.js/next.js/packages/next/dist/export/worker.js:253:36
      Error occurred prerendering page "/500". Read more: https://nextjs.org/docs/messages/prerender-error
      TypeError: ReactDOMServer.renderToReadableStream is not a function
      at Object.renderToInitialStream (../packages/next/dist/server/node-web-streams-helper.js:124:33)
      at renderShell (../packages/next/dist/server/render.js:882:65)
      at Object.renderPage (../packages/next/dist/server/render.js:795:28)
      at Object.defaultGetInitialProps (../packages/next/dist/server/render.js:391:51)
      at Function.getInitialProps (integration/amphtml-custom-optimizer/.next/server/pages/_document.js:590:20)
      at Object.loadGetInitialProps (../packages/next/dist/shared/lib/utils.js:65:29)
      at documentInitialProps (../packages/next/dist/server/render.js:817:48)
      at renderDocument (../packages/next/dist/server/render.js:942:49)
      at renderToHTML (../packages/next/dist/server/render.js:971:34)
      at ../async /home/runner/work/next.js/next.js/packages/next/dist/export/worker.js:253:36
      > Build error occurred
      Error: Export encountered errors on following paths:
      	/
      	/_error: /404
      	/_error: /500
      at ../packages/next/dist/export/index.js:397:19
      at async Span.traceAsyncFn (../packages/next/dist/trace/trace.js:79:20)
      at ../async /home/runner/work/next.js/next.js/packages/next/dist/build/index.js:1012:17
      at async Span.traceAsyncFn (../packages/next/dist/trace/trace.js:79:20)
      at ../async /home/runner/work/next.js/next.js/packages/next/dist/build/index.js:888:13
      at async Span.traceAsyncFn (../packages/next/dist/trace/trace.js:79:20)
      at async Object.build [as default] (../packages/next/dist/build/index.js:57:25)
      info  - Generating static pages (3/3)
      at ChildProcess.<anonymous> (lib/next-test-utils.js:218:11)

  ● AMP Custom Optimizer › should build and start for dynamic page

    command failed with code 1
    warn  - You have enabled experimental feature(s).
    warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

    info  - Checking validity of types...
    warn  - The Next.js plugin was not detected in your ESLint configuration. See https://nextjs.org/docs/basic-features/eslint#migrating-existing-config
    info  - Creating an optimized production build...
    info  - Using locally built binary of @next/swc
    info  - Compiled successfully
    info  - Collecting page data...
    info  - Generating static pages (0/3)

    Error occurred prerendering page "/". Read more: https://nextjs.org/docs/messages/prerender-error
    TypeError: ReactDOMServer.renderToReadableStream is not a function

      122 | }
      123 | async function renderToInitialStream({ ReactDOMServer , element  }) {
    > 124 |     return await ReactDOMServer.renderToReadableStream(element);
          |                                 ^
      125 | }
      126 | async function continueFromInitialStream({ suffix , dataStream , generateStaticHTML , flushEffectHandler , renderStream  }) {
      127 |     const closeTag = '</body></html>';

      at Object.renderToInitialStream (../packages/next/dist/server/node-web-streams-helper.js:124:33)
      at renderShell (../packages/next/dist/server/render.js:882:65)
      at Object.renderPage (../packages/next/dist/server/render.js:795:28)
      at Object.defaultGetInitialProps (../packages/next/dist/server/render.js:391:51)
      at Function.getInitialProps (integration/amphtml-custom-optimizer/.next/server/pages/_document.js:590:20)
      at Object.loadGetInitialProps (../packages/next/dist/shared/lib/utils.js:65:29)
      at documentInitialProps (../packages/next/dist/server/render.js:817:48)
      at renderDocument (../packages/next/dist/server/render.js:942:49)
      at renderToHTML (../packages/next/dist/server/render.js:971:34)
      at ../async /home/runner/work/next.js/next.js/packages/next/dist/export/worker.js:253:36
      Error occurred prerendering page "/404". Read more: https://nextjs.org/docs/messages/prerender-error
      TypeError: ReactDOMServer.renderToReadableStream is not a function
      at Object.renderToInitialStream (../packages/next/dist/server/node-web-streams-helper.js:124:33)
      at renderShell (../packages/next/dist/server/render.js:882:65)
      at Object.renderPage (../packages/next/dist/server/render.js:795:28)
      at Object.defaultGetInitialProps (../packages/next/dist/server/render.js:391:51)
      at Function.getInitialProps (integration/amphtml-custom-optimizer/.next/server/pages/_document.js:590:20)
      at Object.loadGetInitialProps (../packages/next/dist/shared/lib/utils.js:65:29)
      at documentInitialProps (../packages/next/dist/server/render.js:817:48)
      at renderDocument (../packages/next/dist/server/render.js:942:49)
      at renderToHTML (../packages/next/dist/server/render.js:971:34)
      at ../async /home/runner/work/next.js/next.js/packages/next/dist/export/worker.js:253:36
      Error occurred prerendering page "/500". Read more: https://nextjs.org/docs/messages/prerender-error
      TypeError: ReactDOMServer.renderToReadableStream is not a function
      at Object.renderToInitialStream (../packages/next/dist/server/node-web-streams-helper.js:124:33)
      at renderShell (../packages/next/dist/server/render.js:882:65)
      at Object.renderPage (../packages/next/dist/server/render.js:795:28)
      at Object.defaultGetInitialProps (../packages/next/dist/server/render.js:391:51)
      at Function.getInitialProps (integration/amphtml-custom-optimizer/.next/server/pages/_document.js:590:20)
      at Object.loadGetInitialProps (../packages/next/dist/shared/lib/utils.js:65:29)
      at documentInitialProps (../packages/next/dist/server/render.js:817:48)
      at renderDocument (../packages/next/dist/server/render.js:942:49)
      at renderToHTML (../packages/next/dist/server/render.js:971:34)
      at ../async /home/runner/work/next.js/next.js/packages/next/dist/export/worker.js:253:36
      info  - Generating static pages (3/3)
      > Build error occurred
      Error: Export encountered errors on following paths:
      	/
      	/_error: /404
      	/_error: /500
      at ../packages/next/dist/export/index.js:397:19
      at async Span.traceAsyncFn (../packages/next/dist/trace/trace.js:79:20)
      at ../async /home/runner/work/next.js/next.js/packages/next/dist/build/index.js:1012:17
      at async Span.traceAsyncFn (../packages/next/dist/trace/trace.js:79:20)
      at ../async /home/runner/work/next.js/next.js/packages/next/dist/build/index.js:888:13
      at async Span.traceAsyncFn (../packages/next/dist/trace/trace.js:79:20)
      at async Object.build [as default] (../packages/next/dist/build/index.js:57:25)
      at ChildProcess.<anonymous> (lib/next-test-utils.js:218:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| buildDuration | 15.4s | 15s | -328ms |
| buildDurationCached | 6.1s | 6.1s | ⚠️ +67ms |
| nodeModulesSize | 463 MB | 463 MB | -486 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.049 | 3.513 | ⚠️ +0.46 |
| / avg req/sec | 820.06 | 711.6 | ⚠️ -108.46 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.191 | 1.169 | -0.02 |
| /error-in-render avg req/sec | 2099.83 | 2138.83 | +39 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.3 kB | 72.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.08 kB | 3.08 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.71 kB | 5.71 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.64 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.3 kB | 16.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 528 B | 528 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| buildDuration | 17.3s | 17.2s | -153ms |
| buildDurationCached | 6s | 6.1s | ⚠️ +55ms |
| nodeModulesSize | 463 MB | 463 MB | -486 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.088 | 3.5 | ⚠️ +0.41 |
| / avg req/sec | 809.51 | 714.33 | ⚠️ -95.18 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.181 | 1.162 | -0.02 |
| /error-in-render avg req/sec | 2116.02 | 2152.31 | +36.29 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29.1 kB | 29.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.9 kB | 72.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 3.08 kB | 3.08 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.76 kB | 2.76 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.5 kB | 16.5 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js opt-render | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: ea77b1d3a0066464fd6c895a88054e916b6288ee

File: packages/next/build/index.ts [hunk 1]
```diff
@@ -162,7 +162,7 @@ export default async function build(
 
     // We enable concurrent features (Fizz-related rendering architecture) when
     // using React 18 or experimental.
-    const hasReactRoot = shouldUseReactRoot()
+    const hasReactRoot = shouldUseReactRoot
```
Comment (shuding) [inline packages/next/build/index.ts:165]: Why not directly use `process.env.__NEXT_REACT_ROOT` in such places?

**Review by shuding — APPROVED** [review]

**Comment by adhirajsinghchauhan** [comment]
Note: this change causes `Error: Cannot find module 'react-dom/server.browser.js'` when using `@preact/compat`, since it doesn't provide [`server.*.js` files][preact-compat].

[preact-compat]: https://github.com/preactjs/preact/tree/master/compat

**Comment by huozhi** [comment]
@adhirajsinghchauhan I filed #36749 to tweak some changes on next.js side as short term workaround but ideally preact needs to support `/server.(node|browser)` in long term exports which is already exported in react 17. Any library could import from `react-dom/server.node` or `react-dom/server.browser`.

**Pull request merged**